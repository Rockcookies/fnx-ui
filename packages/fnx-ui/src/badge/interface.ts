import { ReactNode, HTMLAttributes } from 'react';

export type BadgeProps = BadgeComponentProps &
	Omit<HTMLAttributes<HTMLSpanElement>, 'color'>;

export interface BadgeComponentProps {
	count?: ReactNode;
	showZero?: boolean;
	dot?: boolean;
	max?: string | number;
	color?: string;
	offset?: [number | string, number | string];
}
