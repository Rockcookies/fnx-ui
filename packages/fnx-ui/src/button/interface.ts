import { ButtonHTMLAttributes, ReactNode } from 'react';

export type ButtonType =
	| 'default'
	| 'primary'
	| 'success'
	| 'warning'
	| 'danger';

export type ButtonSize = 'lg' | 'md' | 'sm' | 'xs';

export type ButtonShape = 'rect' | 'radius' | 'round';

export interface ButtonComponentProps {
	type?: ButtonType;
	size?: ButtonSize;
	icon?: ReactNode;
	iconPosition?: 'left' | 'right';
	loading?: boolean;
	loadingIcon?: ReactNode;
	loadingChildren?: ReactNode;
	disabled?: boolean;
	hairline?: boolean;
	plain?: boolean;
	shape?: ButtonShape;
	block?: boolean;
	color?: string;
	htmlType?: ButtonHTMLAttributes<HTMLButtonElement>['type'];
}

export type ButtonProps = ButtonComponentProps &
	Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'type'>;
