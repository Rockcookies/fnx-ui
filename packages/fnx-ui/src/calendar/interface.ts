import { HTMLAttributes, ReactNode } from 'react';
import CalendarDay from './CalendarDay';
import CalendarMulti from './CalendarMulti';
import CalendarRange from './CalendarRange';

export type CalendarValue = Date | string | number;

export type CalendarDayType =
	| 'normal'
	| 'start'
	| 'start-end'
	| 'middle'
	| 'end'
	| 'selected'
	| 'multiple-middle'
	| 'multiple-selected'
	| 'disabled';

export interface CalendarDayComponentProps {
	color?: string;
	dayHeight?: string;
	date: Date;
	content?: ReactNode;
	type: CalendarDayType;
	topInfo?: ReactNode;
	bottomInfo?: ReactNode;
}

export type CalendarDayProps = CalendarDayComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'content'>;

export interface CalendarSlots {
	day?: (day: CalendarDayComponentProps) => ReactNode;
	subTitle?: boolean | ((date: Date) => ReactNode);
	monthMark?: boolean | ((date: Date) => ReactNode);
	monthTitle?: boolean | ((date: Date) => ReactNode);
	footer?: ReactNode;
}

export interface CalendarBaseProps {
	title?: ReactNode;
	color?: string;
	minDate?: Date | string | number;
	maxDate?: Date | string | number;
	dayHeight?: number | string;
	readonly?: boolean;
	closeIcon?: ReactNode;
	firstDayOfWeek?: number;
	slots?: CalendarSlots;
	confirmText?: ReactNode;
	confirmDisabledText?: ReactNode;
}

export type CalendarElementProps = Omit<
	HTMLAttributes<HTMLDivElement>,
	'defaultValue' | 'onChange' | 'title'
>;

export interface CalendarContainerRef {
	element: HTMLDivElement | null;
	reset: () => void;
	getValue: () => Date[] | undefined;
}

export interface CalendarRef {
	element: HTMLDivElement | null;
	reset: () => void;
	getValue: () => Date | undefined;
}

export interface CalendarComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue;
	onChange?: (value: Date) => void;
	onConfirm?: (value?: Date) => void;
	onCancel?: () => void;
}

export type CalendarProps = CalendarComponentProps & CalendarElementProps;

export type CalendarRangeRef = CalendarContainerRef;

export interface CalendarRangeComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue[];
	onChange?: (value: Date[]) => void;
	onConfirm?: (value?: Date[]) => void;
	onCancel?: () => void;
	allowSameDay?: boolean;
	maxSize?: number;
	onMaxSize?: () => void;
}

export type CalendarRangeProps = CalendarRangeComponentProps &
	CalendarElementProps;

export type CalendarMultiRef = CalendarRangeRef;

export interface CalendarMultiComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue[];
	onChange?: (value: Date[]) => void;
	onConfirm?: (value?: Date[]) => void;
	onCancel?: () => void;
	maxSize?: number;
	onMaxSize?: () => void;
}

export type CalendarMultiProps = CalendarMultiComponentProps &
	CalendarElementProps;

export interface CalendarExportExtra {
	Range: typeof CalendarRange;
	Multi: typeof CalendarMulti;
	Day: typeof CalendarDay;
}
