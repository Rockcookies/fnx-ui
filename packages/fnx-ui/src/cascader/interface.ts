import { HTMLAttributes, ReactNode } from 'react';

export type CascaderValue = string | number;

export interface CascaderOption {
	label?: ReactNode;
	value?: CascaderOption;
	loading?: boolean;
	disabled?: boolean;
	children?: CascaderOption[];
	// for custom filed names
	[key: string]: any;
}

export interface CascaderDataNames {
	label?: string | string[];
	value?: string | string[];
	loading?: string | string[];
	disabled?: string | string[];
	children?: string | string[];
}

export interface CascaderComponentProps<T = CascaderOption> {
	defaultValue?: CascaderValue[];
	data?: T[];
	dataNames?: CascaderDataNames;
	onLoadData?: (options: T[]) => void;
	onConfirm?: (value: CascaderValue[], options: T[]) => void;
	onCancel?: () => void;
	onChange?: (value: CascaderValue[], options: T[]) => void;
	title?: ReactNode;
	placeholder?: ReactNode;
	swipeable?: boolean;
	closeIcon?: ReactNode;
	activeColor?: string;
}

export type CascaderProps<T = CascaderOption> = CascaderComponentProps<T> &
	Omit<HTMLAttributes<HTMLDivElement>, 'defaultValue' | 'title' | 'onChange'>;
