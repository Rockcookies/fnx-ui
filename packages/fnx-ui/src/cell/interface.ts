import { HTMLAttributes, ReactNode } from 'react';
import CellGroup from './CellGroup';

export type CellItemsAlign = 'top' | 'middle' | 'bottom';

export interface CellComponentProps {
	title?: ReactNode;
	titleProps?: HTMLAttributes<HTMLDivElement>;
	content?: ReactNode;
	contentProps?: HTMLAttributes<HTMLDivElement>;
	description?: ReactNode;
	descriptionProps?: HTMLAttributes<HTMLDivElement>;
	leftIcon?: ReactNode;
	leftIconProps?: HTMLAttributes<HTMLSpanElement>;
	rightIcon?: ReactNode;
	rightIconProps?: HTMLAttributes<HTMLSpanElement>;
	border?: boolean;
	itemsAlign?: CellItemsAlign;
	clickable?: boolean;
	size?: 'md' | 'lg';
}

export type CellProps = CellComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

export interface CellGroupComponentProps {
	title?: ReactNode;
	titleProps?: HTMLAttributes<HTMLDivElement>;
	border?: boolean;
	insert?: boolean;
	bodyProps?: HTMLAttributes<HTMLDivElement>;
}

export type CellGroupProps = CellGroupComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

export interface CellExportExtra {
	Group: typeof CellGroup;
}
