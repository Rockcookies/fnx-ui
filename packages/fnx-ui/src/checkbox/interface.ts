import { HTMLAttributes, ReactNode } from 'react';
import InternalCheckboxGroup from './CheckboxGroup';

export type CheckboxValue = string | number | boolean;

export interface CheckboxBaseProps {
	disabled?: boolean;
	direction?: 'vertical' | 'horizontal';
	checkedIcon?: ReactNode;
	icon?: ReactNode;
	iconSize?: string | number;
	iconPosition?: 'left' | 'right';
	iconShape?: 'square' | 'round' | 'plain';
	iconCheckedColor?: string;
	labelDisabled?: boolean;
}

export interface CheckboxComponentProps extends CheckboxBaseProps {
	value?: CheckboxValue;
	checked?: boolean;
	defaultChecked?: boolean;
	onChange?: (checked: boolean) => void;
	skipGroup?: boolean;
}

export type CheckboxProps = CheckboxComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'onChange' | 'value'>;

export interface CheckboxRef {
	element: HTMLDivElement | null;
	toggle: () => void;
}

export interface CheckboxGroupComponentProps<
	T extends CheckboxValue = CheckboxValue,
> extends CheckboxBaseProps {
	value?: T[];
	defaultValue?: T[];
	onChange?: (v: T[]) => void;
	maxCheckedCount?: number;
}

export type CheckboxGroupProps<T extends CheckboxValue = CheckboxValue> =
	CheckboxGroupComponentProps<T> &
		Omit<HTMLAttributes<HTMLDivElement>, 'onChange' | 'defaultValue'>;

export interface CheckboxGroupContextData extends CheckboxBaseProps {
	value?: CheckboxValue[];
	registerValue?: (value: CheckboxValue) => void;
	cancelValue?: (value: CheckboxValue) => void;
	toggleValue?: (isCheck: boolean, value: CheckboxValue) => void;
}
export interface CheckboxGroupRef {
	element: HTMLDivElement | null;
	toggleAll: (options?: { checkAll?: boolean }) => void;
}

export interface CheckboxExportExtra {
	Group: typeof InternalCheckboxGroup;
}
