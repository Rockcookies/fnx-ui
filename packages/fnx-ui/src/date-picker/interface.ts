import { PickerBaseProps, PickerElementProps } from '../picker/interface';
import TimePicker from './TimePicker';

export type DatePickerField = 'year' | 'month' | 'day' | 'hour' | 'minute';

/**TimePicker */
export interface TimePickerComponentProps extends PickerBaseProps {
	defaultValue?: string | number | Date;
	minHour?: number;
	maxHour?: number;
	minMinute?: number;
	maxMinute?: number;
	filter?: (type: DatePickerField, value: number[]) => number[];
	formatter?: (type: DatePickerField, value: string) => string;
	onChange?: (value?: string) => void;
	onConfirm?: (value?: string) => void;
	onCancel?: (value?: string) => void;
}

export type TimePickerProps = TimePickerComponentProps & PickerElementProps;

export interface TimePickerRef {
	root: HTMLDivElement | null;
	getValue: () => string | undefined;
}

/**DateTimePicker */
export interface DatePickerComponentProps extends PickerBaseProps {
	columnsLayout?: string;
	defaultValue?: Date | string | number;
	minDate?: Date | string | number;
	maxDate?: Date | string | number;
	filter?: (type: DatePickerField, value: number[]) => number[];
	formatter?: (type: DatePickerField, value: string) => string;
	onChange?: (value?: Date) => void;
	onConfirm?: (value?: Date) => void;
	onCancel?: (value?: Date) => void;
}

export type DatePickerProps = DatePickerComponentProps & PickerElementProps;

export interface DatePickerRef {
	root: HTMLDivElement | null;
	getValue: () => Date | undefined;
}

export interface DatePickerExportExtra {
	TimePicker: typeof TimePicker;
}
