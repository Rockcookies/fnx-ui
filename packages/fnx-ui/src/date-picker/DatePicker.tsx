import React, {
	forwardRef,
	useCallback,
	useImperativeHandle,
	useMemo,
	useRef,
	useState,
} from 'react';
import configComponentProps from '../hooks/config-component-props';
import useCreation from '../hooks/use-creation';
import useMergedPropRef from '../hooks/use-merged-prop-ref';
import useUpdateEffect from '../hooks/use-update-effect';
import Picker from '../picker/index';
import { PickerMultiRef, PickerOption } from '../picker/interface';
import { padZero } from '../utils/format';
import { Dictionary } from '../utils/interface';
import { isEqualArrays, noop } from '../utils/misc';
import useColumnLayout from './hooks/use-column-layout';
import useDateTime from './hooks/use-datetime';
import { DatePickerField, DatePickerProps, DatePickerRef } from './interface';
import { DateTime, iterateNumber } from './utils';

const useProps = configComponentProps<
	Required<
		Pick<
			DatePickerProps,
			'columnsLayout' | 'filter' | 'formatter' | 'onChange'
		>
	>
>({
	columnsLayout: 'year,month,day',
	filter: (_, v) => v,
	formatter: (_, v) => v,
	onChange: noop,
});

const getMonthEndDay = (year: number, month: number): number => {
	return 32 - new Date(year, month - 1, 32).getDate();
};

const range = (date: DateTime, min: DateTime, max: DateTime): DateTime => {
	const time = date.getTime();

	if (time < min.getTime()) {
		return min;
	}

	if (time > max.getTime()) {
		return max;
	}

	return date;
};

const DatePicker = forwardRef<DatePickerRef, DatePickerProps>((_props, ref) => {
	const [
		props,
		{
			minDate: _minDate,
			maxDate: _maxDate,
			onConfirm,
			onCancel,
			defaultValue: _defaultValue,
			...restProps
		},
	] = useProps(_props);

	const { columnsLayout: _columnsLayout } = props;

	const propsRef = useMergedPropRef(props);
	const columnsLayout = useColumnLayout(_columnsLayout);

	const [defaultMinDate, defaultMaxDate] = useMemo<
		[DateTime, DateTime]
	>(() => {
		const year = new Date().getFullYear();
		return [
			new DateTime(new Date(year - 10, 0, 1)),
			new DateTime(new Date(year + 10, 11, 31)),
		];
	}, []);

	const minDate = useDateTime(_minDate, defaultMinDate);
	const maxDate = useDateTime(_maxDate, defaultMaxDate);

	const [defaultValue, pickerDefaultValue] = useCreation<
		[DateTime, number[]]
	>(() => {
		let dt = new DateTime(
			_defaultValue != null ? _defaultValue : minDate.getDate(),
		);
		dt = range(dt.isValid() ? dt : minDate, minDate, maxDate);

		return [
			dt,
			columnsLayout.map<number>((type) => {
				if (type === 'year') {
					return dt.getYear();
				} else if (type === 'month') {
					return dt.getMonth() + 1;
				} else if (type === 'day') {
					return dt.getDay();
				} else if (type === 'hour') {
					return dt.getHour();
				} else {
					return dt.getMinute();
				}
			}),
		];
	}, []);

	const [value, setValue] = useState<Date>(() => defaultValue.getDate());
	const valueRef = useMergedPropRef(value);

	const pickerRef = useRef<PickerMultiRef>(null);
	const rootRef = useMemo<DatePickerRef>(
		() => ({
			element: pickerRef.current?.element || null,
			getValue: () => valueRef.current,
		}),
		[valueRef],
	);

	useImperativeHandle<DatePickerRef, DatePickerRef>(ref, () => rootRef, [
		rootRef,
	]);

	const getBoundary = useCallback(
		(type: 'min' | 'max', value: Date) => {
			const boundary = type === 'min' ? minDate : maxDate;

			const year = boundary.getYear();
			let month = 1;
			let date = 1;
			let hour = 0;
			let minute = 0;

			if (type === 'max') {
				month = 12;
				date = getMonthEndDay(
					value.getFullYear(),
					value.getMonth() + 1,
				);
				hour = 23;
				minute = 59;
			}

			if (value.getFullYear() === year) {
				month = boundary.getMonth() + 1;
				if (value.getMonth() + 1 === month) {
					date = boundary.getDay();
					if (value.getDate() === date) {
						hour = boundary.getHour();
						if (value.getHours() === hour) {
							minute = boundary.getMinute();
						}
					}
				}
			}

			return {
				year,
				month,
				date,
				hour,
				minute,
			};
		},
		[maxDate, minDate],
	);

	const getColumns = useCallback(
		(
			layout: DatePickerField[],
			value: Date,
			prevColumns: Array<PickerOption[]> = [],
		): Array<PickerOption[]> => {
			const minObj = getBoundary('min', value);
			const maxObj = getBoundary('max', value);

			const fields: Dictionary<[number, number]> = {
				year: [minObj.year, maxObj.year],
				month: [minObj.month, maxObj.month],
				day: [minObj.date, maxObj.date],
				hour: [minObj.hour, maxObj.hour],
				minute: [minObj.minute, maxObj.minute],
			};

			return layout.map<PickerOption[]>((type, idx) => {
				const [min, max] = fields[type];

				let values = iterateNumber(max - min + 1, (idx) => min + idx);

				values = propsRef.current.filter(type, values);

				// 性能优化
				const prev = prevColumns[idx];
				const next = values.map((value) => ({
					value,
					label: propsRef.current.formatter(type, padZero(value)),
				}));

				if (
					isEqualArrays<PickerOption>(
						prev,
						next,
						(a, b) => a.label === b.label && a.value === b.value,
					)
				) {
					return prev;
				}

				return next;
			});
		},
		[getBoundary, propsRef],
	);

	const [columns, setColumns] = useState<Array<PickerOption[]>>(() =>
		getColumns(columnsLayout, value),
	);

	useUpdateEffect(() => {
		const limited = range(new DateTime(valueRef.current), minDate, maxDate);

		setValue((prev) => {
			if (!limited.equals(valueRef.current)) {
				return limited.getDate();
			}

			return prev;
		});
	}, [maxDate, propsRef, valueRef, minDate]);

	useUpdateEffect(() => {
		propsRef.current.onChange(value);
	}, [propsRef, value]);

	useUpdateEffect(() => {
		setColumns((prev) => getColumns(columnsLayout, value, prev));
	}, [columnsLayout, getColumns, value]);

	const handleChange = () => {
		const getValue = (type: DatePickerField, def: number): number => {
			const value = pickerRef.current && pickerRef.current.getValue();

			if (value) {
				for (let i = 0; i < columnsLayout.length; i++) {
					if (columnsLayout[i] === type) {
						const v = value[i];

						if (v != null) {
							return +v;
						}
					}
				}
			}

			return def;
		};

		const year = getValue('year', value.getFullYear());
		const month = getValue('month', value.getMonth() + 1);
		let day = getValue('day', value.getDate());
		const maxDay = getMonthEndDay(year, month);
		day = day > maxDay ? maxDay : day;
		const hour = getValue('hour', value.getHours());
		const minute = getValue('minute', value.getMinutes());

		const next = range(
			new DateTime(new Date(year, month - 1, day, hour, minute)),
			minDate,
			maxDate,
		);

		setValue((prev) => {
			return next.equals(prev) ? prev : next.getDate();
		});
	};

	return (
		<Picker.Multi
			data={columns}
			{...restProps}
			defaultValue={pickerDefaultValue}
			ref={pickerRef}
			onChange={handleChange}
			onConfirm={() => {
				onConfirm && onConfirm(value);
			}}
			onCancel={onCancel}
		></Picker.Multi>
	);
});

DatePicker.displayName = 'DatePicker';

export default DatePicker;
