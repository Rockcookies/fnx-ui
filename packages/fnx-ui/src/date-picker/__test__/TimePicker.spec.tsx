import { act, fireEvent, render } from '@testing-library/react';
import React from 'react';
import TestsDOM from '../../../test/dom';
import TestsReact from '../../../test/react';
import DatePicker from '../index';
import { TimePickerProps, TimePickerRef } from '../interface';

describe('<TimePicker/>', () => {
	const renderPicker = (props: TimePickerProps = {}) => {
		const ref = React.createRef<TimePickerRef>();

		const { container, getByTestId, rerender } = render(
			<DatePicker.TimePicker ref={ref} data-testid="picker" {...props} />,
		);

		const confirm = TestsDOM.mustQuerySelector(
			container,
			'.fnx-picker__confirm',
		);

		const cancel = TestsDOM.mustQuerySelector(
			container,
			'.fnx-picker__cancel',
		);

		return {
			container,
			getByTestId,
			rerender,
			ref,
			confirm,
			cancel,
		};
	};

	const filter = (type: string, options: number[]) => {
		const mod = type === 'minute' ? 10 : 5;
		return options.filter((option) => option % mod === 0);
	};

	function formatter(type: string, value: string) {
		if (type === 'month') {
			return `${value}æœˆ`;
		}

		return `${value} ${type}`;
	}

	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		jest.useRealTimers();
	});

	it('format initial value and trigger confirm and cancel event', () => {
		const onConfirm = jest.fn();
		const onCancel = jest.fn();

		const { confirm, cancel } = renderPicker({
			defaultValue: '11:59',
			minHour: 14,
			maxMinute: 50,
			onConfirm,
			onCancel,
		});

		act(() => {
			fireEvent.click(confirm);
			fireEvent.click(cancel);
		});

		expect(onConfirm).toHaveBeenCalledWith('14:00');
		expect(onCancel).toHaveBeenCalled();
	});

	it('filter and formatter prop', () => {
		const { container } = renderPicker({
			filter,
			formatter,
		});

		expect(container).toMatchSnapshot();
	});

	it('change event', async () => {
		const onChange = jest.fn();

		const { container } = renderPicker({
			defaultValue: '12:00',
			onChange,
		});

		const track = container.querySelectorAll('.fnx-picker-column');
		const columns = track[0].querySelectorAll('.fnx-picker-column__option');

		act(() => {
			fireEvent.click(columns[14]);
			jest.advanceTimersByTime(300);
		});

		expect(onChange).toHaveBeenCalledWith('14:00');
	});

	it('get value by timePickerRef', async () => {
		const defaultValue = new Date();
		defaultValue.setHours(10);
		defaultValue.setMinutes(6);

		const { ref, rerender } = renderPicker({
			defaultValue,
		});

		expect(TestsReact.mustGetRef(ref).getValue()).toEqual('10:06');

		rerender(<DatePicker.TimePicker ref={ref} maxHour={9} />);

		expect(TestsReact.mustGetRef(ref).getValue()).toEqual('09:06');
	});
});
