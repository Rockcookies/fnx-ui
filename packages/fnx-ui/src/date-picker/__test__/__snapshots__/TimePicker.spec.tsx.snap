// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimePicker/> filter and formatter prop 1`] = `
<div>
  <div
    class="fnx-picker"
    data-testid="picker"
  >
    <div
      class="fnx-picker__toolbar"
    >
      <button
        class="fnx-picker__cancel"
        type="button"
      >
        Cancel
      </button>
      <button
        class="fnx-picker__confirm"
        type="button"
      >
        Confirm
      </button>
    </div>
    <div
      class="fnx-picker__columns"
      style="height: 264px;"
    >
      <div
        class="fnx-picker-column"
      >
        <ul
          class="fnx-picker-column__wrapper"
          style="transform: translate3d(0, 110px, 0);"
        >
          <li
            class="fnx-picker-column__option fnx-picker-column__option--selected"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="fnx-ellipsis"
            >
              00 hour
            </div>
          </li>
          <li
            class="fnx-picker-column__option"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="fnx-ellipsis"
            >
              05 hour
            </div>
          </li>
          <li
            class="fnx-picker-column__option"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="fnx-ellipsis"
            >
              10 hour
            </div>
          </li>
          <li
            class="fnx-picker-column__option"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="fnx-ellipsis"
            >
              15 hour
            </div>
          </li>
          <li
            class="fnx-picker-column__option"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="fnx-ellipsis"
            >
              20 hour
            </div>
          </li>
        </ul>
      </div>
      <div
        class="fnx-picker-column"
      >
        <ul
          class="fnx-picker-column__wrapper"
          style="transform: translate3d(0, 110px, 0);"
        >
          <li
            class="fnx-picker-column__option fnx-picker-column__option--selected"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="fnx-ellipsis"
            >
              00 minute
            </div>
          </li>
          <li
            class="fnx-picker-column__option"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="fnx-ellipsis"
            >
              10 minute
            </div>
          </li>
          <li
            class="fnx-picker-column__option"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="fnx-ellipsis"
            >
              20 minute
            </div>
          </li>
          <li
            class="fnx-picker-column__option"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="fnx-ellipsis"
            >
              30 minute
            </div>
          </li>
          <li
            class="fnx-picker-column__option"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="fnx-ellipsis"
            >
              40 minute
            </div>
          </li>
          <li
            class="fnx-picker-column__option"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="fnx-ellipsis"
            >
              50 minute
            </div>
          </li>
        </ul>
      </div>
      <div
        class="fnx-picker__mask"
        style="background-size: 100% 110px;"
      />
      <div
        class="fnx-picker__frame fnx-hairline-unset--top-bottom"
        style="height: 44px;"
      />
    </div>
  </div>
</div>
`;
